_posts/2020-11-23-how-fast-are-ractors.md

```diff
@@ -0,0 +1,133 @@
+
+So I [wrote a benchmark](https://github.com/noahgibbs/ractor_basic_benchmarks) with some CPU time to it. Specifically, I wrote three benchmarks &mdash; one with threads, one multiprocess and one with ractors &mdash; that each calculate whether each number in a 100-number chunk is prime. Then the 100 prime-ness calculations are turned into a bit vector with the first numbers as the highest-order bits. Here's the code:
+
+```
+bools = (item..(item+99)).map { |nn| nn.prime? }
+p_int = bools.inject(0) { |total, b| total * 2 + (b ? 1 : 0) }
+# return [item, p_int]
+```
+
+The prime calculation uses Ruby's Prime library, which is very simple and written in Ruby. You can [read the source](https://github.com/ruby/ruby/tree/master/lib/prime.rb) if you like. It doesn't do anything fancy, it doesn't release the GVL, and it calculates heavily in Ruby.
+
+All three benchmark types (fork, threads, Ractors) calculate the same things in the same way.
+
+The "fork" benchmark runs as a supervisor process, spawns N worker processes and talks to them over interprocess pipes. The "thread" benchmark uses a Mutex and a Ruby array as a queue. And the "ractor" benchmark uses [send/recv and yield/take](https://github.com/ruby/ruby/blob/master/doc/ractor.md) to pass results in the appropriate Ractor style.
+
+All results are sent back, sorted and checksummed to make sure all three benchmarks agree. The benchmark gets to time itself to see how long just the actual processing takes, and the [benchmark coordinator](https://github.com/noahgibbs/ractor_basic_benchmarks/tree/main/comparison_collector.rb) also times each process to see how long it takes with full setup and shutdown. The benchmarks are also run in a randomised order to keep varying load on the host-system load from turning into bias.
+
+And I originally timed this ***on my MacBook. Here's a tip: don't***. The performance and stability are substantially worse on Mac. I was getting around a 20% crash-and-error rate for medium-size benchmarks and some ***weird*** unexpected results. While things will probably be better by the release date, right now the Mac is ***not*** good for benchmarking Ractors.
+
+I'm still checking this on an itty bitty instance, a Linode 4GB. This is a very prerelease feature. And based on the results I'm seeing, it's not ready for larger-scale benchmarking yet. That's okay, let's have a look anyway.
```

Comment:
@noahgibbs Could you share how many cores were available in this instance?

Reply:
Sure. It's a two-core instance. I added a reference to that here.